src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -473,6 +503,77 @@ public void RemoveAdditionalFile(string fullPath)
 
         #endregion
 
+        #region Non Source File Addition/Removal
+        public void AddDynamicSourceFile(string dynamicFilePath, ImmutableArray<string> folders)
+        {
+            var extension = FileNameUtilities.GetExtension(dynamicFilePath)?.TrimStart('.');
+            if (extension?.Length == 0)
+            {
+                return;
+            }
+
+            foreach (var provider in _dynamicFileInfoProviders)
+            {
+                // skip unrelated providers
+                if (!provider.Metadata.Extensions.Any(e => string.Equals(e, extension, StringComparison.OrdinalIgnoreCase)))
+                {
+                    continue;
+                }
+
+                // don't get confused by _filePath and filePath
+                // VisualStudioProject._filePath points to csproj/vbproj of the project
+                // and the parameter filePath points to dynamic file such as cshtml and etc
+                // 
+                // also, provider is free-threaded. so find to call Wait rather than JTF
+                var fileInfo = provider.Value.GetDynamicFileInfoAsync(
+                    projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
+
+                if (fileInfo == null)
+                {
+                    continue;
+                }
+
+                fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
+
+                // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
+                _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
```

Comment:
instead of neeidng to keep these separate maps, why not just instantiate a type we own that holds all the data we need, then pass that to .AddDynamicFile?

Reply:
I dont know how that will look like but sure we can do such thing but not in this PR.

Reply:
Could you clarify why not in this PR?  This is the PR that is adding all this new code and capabilities.  Why not have that initial implementation follow these sorts of good design approaches?  

Note: it feels lke this isn't an actual PR as these concerns about quality and actual implementation appropriateness are not being addressed.